File: rust/src/scalar_value.rs
use automerge as am;

pub enum ScalarValue {
    Bytes { value: Vec<u8> },
    String { value: String },
    Int { value: i64 },
    Uint { value: u64 },
    F64 { value: f64 },
    Counter { value: i64 },
    Timestamp { value: i64 },
    Boolean { value: bool },
    Unknown { type_code: u8, data: Vec<u8> },
    Null,
}

impl From<ScalarValue> for am::ScalarValue {
    fn from(value: ScalarValue) -> Self {
        match value {
            ScalarValue::Bytes { value } => am::ScalarValue::Bytes(value),
            ScalarValue::String { value } => am::ScalarValue::Str(value.into()),
            ScalarValue::Int { value } => am::ScalarValue::Int(value),
            ScalarValue::Uint { value } => am::ScalarValue::Uint(value),
            ScalarValue::F64 { value } => am::ScalarValue::F64(value),
            ScalarValue::Counter { value } => am::ScalarValue::Counter(value.into()),
            ScalarValue::Timestamp { value } => am::ScalarValue::Timestamp(value),
            ScalarValue::Boolean { value } => am::ScalarValue::Boolean(value),
            ScalarValue::Unknown { type_code, data } => am::ScalarValue::Unknown {
                type_code,
                bytes: data,
            },
            ScalarValue::Null => am::ScalarValue::Null,
        }
    }
}

impl<'a> From<&'a am::ScalarValue> for ScalarValue {
    fn from(value: &'a am::ScalarValue) -> Self {
        match value {
            am::ScalarValue::Bytes(b) => ScalarValue::Bytes { value: b.clone() },
            am::ScalarValue::Str(s) => ScalarValue::String {
                value: s.to_string(),
            },
            am::ScalarValue::Int(i) => ScalarValue::Int { value: *i },
            am::ScalarValue::Uint(u) => ScalarValue::Uint { value: *u },
            am::ScalarValue::F64(f) => ScalarValue::F64 { value: *f },
            am::ScalarValue::Counter(i) => ScalarValue::Counter { value: i.into() },
            am::ScalarValue::Timestamp(i) => ScalarValue::Timestamp { value: *i },
            am::ScalarValue::Boolean(b) => ScalarValue::Boolean { value: *b },
            am::ScalarValue::Unknown { type_code, bytes } => ScalarValue::Unknown {
                type_code: *type_code,
                data: bytes.clone(),
            },
            am::ScalarValue::Null => ScalarValue::Null,
        }
    }
}

File: Sources/Automerge/ScalarValue.swift
import enum AutomergeUniffi.ScalarValue
import Foundation

typealias FFIScalar = AutomergeUniffi.ScalarValue

/// A type that represents a primitive Automerge value.
public enum ScalarValue: Equatable, Hashable, Sendable {
    /// A byte buffer.
    case Bytes(Data)
    /// A string.
    case String(String)
    /// An unsigned integer.
    case Uint(UInt64)
    /// A signed integer.
    case Int(Int64)
    /// A floating point number.
    case F64(Double)
    /// An integer counter.
    case Counter(Int64)
    /// A 64-bit signed integer that represents the number of milliseconds since the UNIX epoch.
    case Timestamp(Date)
    /// A Boolean value.
    case Boolean(Bool)
    /// An unknown, raw scalar type.
    ///
    /// This type is reserved for forward compatibility, and is not expected to be created directly.
    case Unknown(typeCode: UInt8, data: Data)
    /// A null.
    case Null

    func toFfi() -> FFIScalar {
        switch self {
        case let .Bytes(b):
            return .bytes(value: Array(b))
        case let .String(s):
            return .string(value: s)
        case let .Uint(i):
            return .uint(value: i)
        case let .Int(i):
            return .int(value: i)
        case let .F64(d):
            return .f64(value: d)
        case let .Counter(i):
            return .counter(value: i)
        case let .Timestamp(date):
            return .timestamp(value: Int64(date.timeIntervalSince1970 * 1000))
        case let .Boolean(v):
            return .boolean(value: v)
        case let .Unknown(t, d):
            return .unknown(typeCode: t, data: Array(d))
        case .Null:
            return .null
        }
    }

    static func fromFfi(value: FFIScalar) -> Self {
        switch value {
        case let .bytes(value):
            return .Bytes(Data(value))
        case let .string(value):
            return .String(value)
        case let .uint(value):
            return .Uint(value)
        case let .int(value):
            return .Int(value)
        case let .f64(value):
            return .F64(value)
        case let .counter(value):
            return .Counter(value)
        case let .timestamp(value):
            return .Timestamp(Date(timeIntervalSince1970: Double(value) / 1000))
        case let .boolean(value):
            return .Boolean(value)
        case let .unknown(typeCode, data):
            return .Unknown(typeCode: typeCode, data: Data(data))
        case .null:
            return .Null
        }
    }
}

extension ScalarValue: CustomStringConvertible {
    public var description: String {
        switch self {
        case let .Boolean(boolValue):
            return "Boolean(\(boolValue))"
        case let .Bytes(data):
            var stringRep = "Data("
            if data.count > 16 {
                let first16Bytes = data[0 ..< 15]
                stringRep.append(first16Bytes.map { Swift.String(format: "%02hhx", $0) }.joined())
            } else {
                stringRep.append(data.map { Swift.String(format: "%02hhx", $0) }.joined())
            }
            return stringRep.appending(")")
        case let .String(stringVal):
            return "String(\(stringVal))"
        case let .Uint(uintVal):
            return "UInt(\(uintVal))"
        case let .Int(intValue):
            return "Int(\(intValue))"
        case let .F64(doubleValue):
            #if canImport(Darwin)
            if #available(iOS 15.0, macOS 12.0, *) {
                return "Double(\(doubleValue.formatted(.number.precision(.significantDigits(2)))))"
            } else {
                return "Double(\(doubleValue))"
            }
            #else
            // swift-corelibs-foundation lacks `formatted` support
            return "Double(\(doubleValue))"
            #endif
        case let .Counter(intValue):
            return "Counter(\(intValue))"
        case let .Timestamp(intValue):
            return "Timestamp(\(intValue))"
        case let .Unknown(typeCode: typeCode, data: data):
            return "Unknown(type: \(typeCode), data: \(data.map { Swift.String(format: "%02hhx", $0) }.joined()))"
        case .Null:
            return "Null()"
        }
    }
}
